endpoints/openrtb2/video_auction.go

```diff
@@ -0,0 +1,299 @@
+}
+
+/*
+1. Parse "storedrequestid" field from simplified endpoint request body.
+2. If config flag to require that field is set (which it will be for us) and this field is not given then error out here.
+3. Load the stored request JSON for the given storedrequestid, if the id was invalid then error out here.
+4. Use "json-patch" 3rd party library to merge the request body JSON data into the stored request JSON data.
+5. Unmarshal the merged JSON data into a Go structure.
+6. Call setFieldsImplicitly from auction.go to get basic data from the HTTP request into an OpenRTB bid request to start building the OpenRTB bid request.
+7. Add fields from merged JSON data that correspond to an OpenRTB request into the OpenRTB bid request we are building.
+	a. Unmarshal certain OpenRTB defined structs directly into the OpenRTB bid request.
+	b. In cases where customized logic is needed just copy/fill the fields in directly.
+8. Loop through ad pods to build array of Imps into OpenRTB request, for each pod:
+	a. Load the stored impression to use as the basis for impressions generated for this pod from the configid field.
+	b. NumImps = adpoddurationsec / MIN_VALUE(allowedDurations)
+	c. Build impression array for this pod:
+		i. Create array of NumImps entries initialized to the base impression loaded from the configid.
+			1. If requireexactdurations = true, iterate over allowdDurations and for (NumImps / len(allowedDurations)) number of Imps set minduration = maxduration = allowedDurations[i]
+			2. If requireexactdurations = false, set maxduration = MAX_VALUE(allowedDurations)
+		ii. Set Imp.id field to "podX_Y" where X is the pod index and Y is the impression index within this pod.
+	d. Append impressions for this pod to the overall list of impressions in the OpenRTB bid request.
+9. Call validateRequest() function from auction.go to validate the generated request.
+10. Build response in proper format
+*/
+func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+
+	lr := &io.LimitedReader{
+		R: r.Body,
+		N: deps.cfg.MaxRequestSize,
+	}
+	requestJson, err := ioutil.ReadAll(lr)
+	if err != nil {
+		return
+	}
+
+	//load additional data - stored simplified req
+	storedRequestId, err := getVideoStoredRequestId(requestJson)
+	if err != nil {
+		return
+	}
+	storedRequest, err := loadStoredVideoRequest(storedRequestId)
+	if err != nil {
+		return
+	}
+
+	//merge incoming req with stored video req
+	resolvedRequest, err := jsonpatch.MergePatch(requestJson, storedRequest)
+
+	//unmarshal and validate combined result
+	videoBidReq, errl := deps.parseVideoRequest(resolvedRequest)
+	if len(errl) > 0 {
+		return
+	}
+
+	var bidReq = openrtb.BidRequest{}
+
+	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
+	deps.setFieldsImplicitly(r, &bidReq)
```

Comment:
Hans was pointing out that we should swap the order of steps 6 and 7 in that write-up, so we should call setFieldsImplicitly after the mergeData step. That function will only overwrite fields that haven't already been set, so it could result in less overall work if we call it after the merge.

Reply:
Done